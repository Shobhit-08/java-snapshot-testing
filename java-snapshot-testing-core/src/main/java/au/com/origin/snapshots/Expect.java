package au.com.origin.snapshots;

import au.com.origin.snapshots.comparators.SnapshotComparator;
import au.com.origin.snapshots.reporters.SnapshotReporter;
import au.com.origin.snapshots.serializers.SnapshotSerializer;
import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;

@RequiredArgsConstructor
public class Expect {
  private final SnapshotVerifier snapshotVerifier;
  private final Method testMethod;

  private SnapshotSerializer snapshotSerializer;
  private SnapshotComparator snapshotComparator;
  private List<SnapshotReporter> snapshotReporters;
  private String scenario;

  private final Map<String, String> headers = new HashMap<>();

  public static Expect of(SnapshotVerifier snapshotVerifier, Method method) {
    return new Expect(snapshotVerifier, method);
  }

  /**
   * Make an assertion on the given input parameters against what already exists
   *
   * <p>If you were previously using varargs and see an error - you can fix the error using
   * "toMatchSnapshotLegacy", however, a better approach is to use the ".scenario()" feature as
   * future versions of this library will most likely remove the legacy implementation completely.
   *
   * @param object snapshot object
   */
  public void toMatchSnapshot(Object object) {
    SnapshotContext snapshotContext = snapshotVerifier.expectCondition(testMethod, object);
    if (snapshotSerializer != null) {
      snapshotContext.setSnapshotSerializer(snapshotSerializer);
    }
    if (snapshotComparator != null) {
      snapshotContext.setSnapshotComparator(snapshotComparator);
    }
    if (snapshotReporters != null) {
      snapshotContext.setSnapshotReporters(snapshotReporters);
    }
    if (scenario != null) {
      snapshotContext.setScenario(scenario);
    }
    snapshotContext.header.putAll(headers);

    snapshotContext.checkValidContext();

    snapshotContext.toMatchSnapshot();
  }

  /**
   * Normally a snapshot can be applied only once to a test method.
   *
   * <p>For Parameterized tests where the same method is executed multiple times you can supply the
   * scenario() to overcome this restriction. Ensure each scenario is unique.
   *
   * @param scenario - unique scenario description
   * @return Snapshot
   */
  public Expect scenario(String scenario) {
    this.scenario = scenario;
    return this;
  }

  /**
   * Apply a custom serializer for this snapshot
   *
   * @param serializer your custom serializer
   * @return Snapshot
   */
  public Expect serializer(SnapshotSerializer serializer) {
    this.snapshotSerializer = serializer;
    return this;
  }

  /**
   * Apply a custom serializer for this snapshot
   *
   * @param name - the {name} attribute serializer.{name} from snapshot.properties
   * @return Snapshot
   */
  public Expect serializer(String name) {
    this.snapshotSerializer = SnapshotProperties.getInstance("serializer." + name);
    return this;
  }

  /**
   * Apply a custom comparator for this snapshot
   *
   * @param comparator your custom comparator
   * @return Snapshot
   */
  public Expect comparator(SnapshotComparator comparator) {
    this.snapshotComparator = comparator;
    return this;
  }

  /**
   * Apply a custom comparator for this snapshot
   *
   * @param name the {name} attribute comparator.{name} from snapshot.properties
   * @return Snapshot
   */
  public Expect comparator(String name) {
    this.snapshotComparator = SnapshotProperties.getInstance("comparator." + name);
    return this;
  }

  /**
   * Apply a list of custom reporters for this snapshot This will replace the default reporters
   * defined in the config
   *
   * @param reporters your custom reporters
   * @return Snapshot
   */
  public Expect reporters(SnapshotReporter... reporters) {
    this.snapshotReporters = Arrays.asList(reporters);
    return this;
  }

  /**
   * Apply a list of custom reporters for this snapshot This will replace the default reporters
   * defined in the config
   *
   * @param name the {name} attribute reporters.{name} from snapshot.properties
   * @return Snapshot
   */
  public Expect reporters(String name) {
    this.snapshotReporters = SnapshotProperties.getInstances("reporters." + name);
    return this;
  }

  /**
   * Apply a custom serializer for this snapshot.
   *
   * @param serializer your custom serializer
   * @return this
   * @see au.com.origin.snapshots.serializers.SnapshotSerializer
   *     <p>Example implementations
   * @see au.com.origin.snapshots.serializers.ToStringSnapshotSerializer
   * @see au.com.origin.snapshots.serializers.Base64SnapshotSerializer
   */
  @SneakyThrows
  public Expect serializer(Class<? extends SnapshotSerializer> serializer) {
    this.snapshotSerializer = serializer.getConstructor().newInstance();
    return this;
  }

  /**
   * Add anything you like to the snapshot header.
   *
   * <p>These custom headers can be used in serializers, comparators or reporters to change how they
   * behave.
   *
   * @param key key
   * @param value value
   * @return Expect
   */
  public Expect header(String key, String value) {
    headers.put(key, value);
    return this;
  }
}
