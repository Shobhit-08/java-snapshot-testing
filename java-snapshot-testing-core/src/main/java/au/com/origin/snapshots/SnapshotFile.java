package au.com.origin.snapshots;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardOpenOption;
import java.util.Collections;
import java.util.List;
import java.util.Objects;
import java.util.Set;
import java.util.TreeSet;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import lombok.Getter;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;

@Slf4j
public class SnapshotFile {

  public static final String SPLIT_STRING = "\n\n\n";

  private final String fileName;
  private final Class<?> testClass;
  @Getter private Set<Snapshot> snapshots = Collections.synchronizedSortedSet(new TreeSet<>());
  private Set<Snapshot> debugSnapshots = Collections.synchronizedSortedSet(new TreeSet<>());

  SnapshotFile(String srcDirPath, String fileName, Class<?> testClass) throws IOException {
    this.testClass = testClass;
    this.fileName = srcDirPath + File.separator + fileName;
    log.info("Snapshot File: " + this.fileName);

    StringBuilder fileContent = new StringBuilder();

    try (BufferedReader br =
        new BufferedReader(
            new InputStreamReader(new FileInputStream(this.fileName), StandardCharsets.UTF_8))) {

      String sCurrentLine;

      while ((sCurrentLine = br.readLine()) != null) {
        fileContent.append(sCurrentLine + "\n");
      }

      String fileText = fileContent.toString();
      if (!"".equals(fileText.trim())) {
        snapshots =
            Collections.synchronizedSortedSet(
                Stream.of(fileContent.toString().split(SPLIT_STRING))
                    .map(String::trim)
                    .map(Snapshot::parse)
                    .collect(Collectors.toCollection(TreeSet::new)));
      }
    } catch (IOException e) {
      // ...
    }

    deleteDebugFile();
  }

  private String getDebugFilename() {
    return this.fileName + ".debug";
  }

  public File createDebugFile(Snapshot snapshot) {
    File file = null;
    try {
      file = new File(getDebugFilename());
      file.getParentFile().mkdirs();
      file.createNewFile();

      try (FileOutputStream fileStream = new FileOutputStream(file, false)) {
        fileStream.write(snapshot.raw().getBytes(StandardCharsets.UTF_8));
      } catch (IOException e) {
        e.printStackTrace();
      }
    } catch (IOException e) {
      e.printStackTrace();
    }

    return file;
  }

  @SneakyThrows
  public void deleteDebugFile() {
    Files.deleteIfExists(Paths.get(getDebugFilename()));
  }

  @SneakyThrows
  public void delete() {
    Files.deleteIfExists(Paths.get(this.fileName));
  }

  @SneakyThrows
  public synchronized File createFileIfNotExists(String filename) {
    Path path = Paths.get(filename);
    if (!Files.exists(path)) {
      Files.createDirectories(path.getParent());
      Files.createFile(path);
    }
    return path.toFile();
  }

  public void pushSnapshot(Snapshot snapshot) {
    synchronized (snapshots) {
      snapshots.add(snapshot);
      TreeSet<String> rawSnapshots =
          snapshots.stream().map(Snapshot::raw).collect(Collectors.toCollection(TreeSet::new));
      updateFile(this.fileName, rawSnapshots);
    }
  }

  public synchronized void pushDebugSnapshot(Snapshot snapshot) {
    debugSnapshots.add(snapshot);
    TreeSet<String> rawDebugSnapshots =
        debugSnapshots.stream().map(Snapshot::raw).collect(Collectors.toCollection(TreeSet::new));
    updateFile(getDebugFilename(), rawDebugSnapshots);
  }

  private void updateFile(String fileName, Set<String> rawSnapshots) {
    File file = createFileIfNotExists(fileName);
    try (FileOutputStream fileStream = new FileOutputStream(file, false)) {
      byte[] myBytes = String.join(SPLIT_STRING, rawSnapshots).getBytes(StandardCharsets.UTF_8);
      fileStream.write(myBytes);
    } catch (IOException e) {
      e.printStackTrace();
    }
  }

  @SneakyThrows
  public void cleanup() {
    Path path = Paths.get(this.fileName);
    if (Files.exists(path)) {
      if (Files.size(path) == 0 || snapshotsAreTheSame()) {
        deleteDebugFile();
      }

      if (Files.size(path) == 0) {
        delete();
      } else {
        String content =
            new String(Files.readAllBytes(Paths.get(this.fileName)), StandardCharsets.UTF_8);
        Files.write(
            path, content.getBytes(StandardCharsets.UTF_8), StandardOpenOption.TRUNCATE_EXISTING);
      }
    }
  }

  @SneakyThrows
  private boolean snapshotsAreTheSame() {
    Path path = Paths.get(this.getDebugFilename());
    if (Files.exists(path)) {
      List<String> snapshotFileContent =
          Files.readAllLines(Paths.get(this.fileName), StandardCharsets.UTF_8);
      List<String> debugSnapshotFileContent =
          Files.readAllLines(Paths.get(this.getDebugFilename()), StandardCharsets.UTF_8);
      return Objects.equals(snapshotFileContent, debugSnapshotFileContent);
    }

    return false;
  }
}
