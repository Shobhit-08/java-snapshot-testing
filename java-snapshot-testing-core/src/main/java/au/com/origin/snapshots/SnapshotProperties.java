package au.com.origin.snapshots;

import au.com.origin.snapshots.exceptions.MissingSnapshotPropertiesKeyException;
import java.io.InputStream;
import java.util.Arrays;
import java.util.List;
import java.util.Properties;
import java.util.stream.Collectors;
import lombok.extern.slf4j.Slf4j;

@Slf4j
public enum SnapshotProperties {
  INSTANCE;

  Properties snapshotProperties = new Properties();

  SnapshotProperties() {
    try {
      InputStream in = SnapshotProperties.class.getResourceAsStream("/snapshot.properties");
      snapshotProperties.load(in);
    } catch (Exception e) {
      // It's ok, if the SnapshotConfig implementation attempts to get a property they will receive
      // a MissingSnapshotPropertiesKeyException
    }
  }

  public static String getOrThrow(String key) {
    Object value = INSTANCE.snapshotProperties.get(key);
    if (value == null) {
      throw new MissingSnapshotPropertiesKeyException(key);
    }
    return value.toString();
  }

  public static <T> T getInstance(String key) {
    String value = SnapshotProperties.getOrThrow(key);
    return createInstance(value);
  }

  public static <T> List<T> getInstances(String key) {
    String value = SnapshotProperties.getOrThrow(key);
    return Arrays.stream(value.split(","))
        .map(String::trim)
        .map(it -> (T) createInstance(it))
        .collect(Collectors.toList());
  }

  @SuppressWarnings("unchecked")
  private static <T> T createInstance(String className) {
    try {
      Class<?> clazz = Class.forName(className);
      return (T) clazz.newInstance();
    } catch (Exception e) {
      throw new RuntimeException("Unable to instantiate class " + className, e);
    }
  }
}
