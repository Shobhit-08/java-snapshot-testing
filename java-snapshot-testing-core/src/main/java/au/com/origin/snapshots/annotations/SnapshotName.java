package au.com.origin.snapshots.annotations;

import java.lang.annotation.*;

@Documented
@Target({ElementType.METHOD})
@Inherited
@Retention(RetentionPolicy.RUNTIME)
public @interface SnapshotName {
  String value();
}
