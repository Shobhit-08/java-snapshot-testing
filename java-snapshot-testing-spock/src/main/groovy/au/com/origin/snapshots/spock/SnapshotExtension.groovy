package au.com.origin.snapshots.spock

import au.com.origin.snapshots.SnapshotVerifier
import au.com.origin.snapshots.config.PropertyResolvingSnapshotConfig
import au.com.origin.snapshots.config.SnapshotConfig
import au.com.origin.snapshots.config.SnapshotConfigInjector
import org.spockframework.runtime.extension.AbstractAnnotationDrivenExtension
import org.spockframework.runtime.model.SpecInfo

class SnapshotExtension extends AbstractAnnotationDrivenExtension<EnableSnapshots> implements SnapshotConfigInjector {

    SnapshotVerifier snapshotVerifier;

    void visitSpecAnnotation(EnableSnapshots annotation, SpecInfo spec) {
        this.snapshotVerifier = new SnapshotVerifier(getSnapshotConfig(), spec.reflection, false)
    }

    void visitSpec(SpecInfo spec) {
        def snapshotMethodInterceptor = new SnapshotMethodInterceptor(snapshotVerifier)
        spec.allFeatures.featureMethod*.addInterceptor(snapshotMethodInterceptor)
        spec.addCleanupSpecInterceptor(snapshotMethodInterceptor)
    }

    @Override
    SnapshotConfig getSnapshotConfig() {
        return new PropertyResolvingSnapshotConfig()
    }
}