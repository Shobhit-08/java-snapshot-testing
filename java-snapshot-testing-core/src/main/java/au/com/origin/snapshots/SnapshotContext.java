package au.com.origin.snapshots;

import au.com.origin.snapshots.annotations.SnapshotName;
import au.com.origin.snapshots.comparators.SnapshotComparator;
import au.com.origin.snapshots.config.SnapshotConfig;
import au.com.origin.snapshots.exceptions.ReservedWordException;
import au.com.origin.snapshots.exceptions.SnapshotExtensionException;
import au.com.origin.snapshots.exceptions.SnapshotMatchException;
import au.com.origin.snapshots.reporters.SnapshotReporter;
import au.com.origin.snapshots.serializers.SnapshotSerializer;

import java.lang.reflect.Method;
import java.util.*;
import java.util.stream.Collectors;

import au.com.origin.snapshots.util.Constants;
import lombok.Getter;
import lombok.Setter;
import lombok.extern.slf4j.Slf4j;

@Slf4j
public class SnapshotContext {

    private static final List<String> RESERVED_WORDS = Arrays.asList("=", "[", "]");

    private final SnapshotConfig snapshotConfig;
    private final SnapshotFile snapshotFile;

    @Getter
    final Class<?> testClass;

    @Getter
    final Method testMethod;

    final Object current;
    private final boolean isCI;

    @Setter
    private SnapshotSerializer snapshotSerializer;
    @Setter
    private SnapshotComparator snapshotComparator;
    @Setter
    private List<SnapshotReporter> snapshotReporters;

    @Setter
    @Getter
    String scenario;

    @Getter
    SnapshotHeader header = new SnapshotHeader();

    SnapshotContext(
            SnapshotConfig snapshotConfig,
            SnapshotFile snapshotFile,
            Class<?> testClass,
            Method testMethod,
            Object current) {
        this.snapshotConfig = snapshotConfig;
        this.snapshotFile = snapshotFile;
        this.testClass = testClass;
        this.testMethod = testMethod;
        this.current = current;

        this.isCI = snapshotConfig.isCI();
        this.snapshotSerializer = snapshotConfig.getSerializer();
        this.snapshotComparator = snapshotConfig.getComparator();
        this.snapshotReporters = snapshotConfig.getReporters();
        this.scenario = null;
    }

    public void toMatchSnapshot() {

        Set<Snapshot> rawSnapshots = snapshotFile.getSnapshots();
        Snapshot previousSnapshot = getRawSnapshot(rawSnapshots);
        Snapshot currentSnapshot = takeSnapshot();

        if (previousSnapshot != null && shouldUpdateSnapshot()) {
            snapshotFile.getSnapshots().remove(previousSnapshot);
            previousSnapshot = null;
        }

        if (previousSnapshot != null) {
            // generate debug files only when not running in shadowMode
            if (System.getProperty(Constants.SHADOW_MODE) == null || "false".equals(System.getProperty(Constants.SHADOW_MODE))) {
                snapshotFile.pushDebugSnapshot(currentSnapshot);

                // Match existing Snapshot
                if (!snapshotComparator.matches(previousSnapshot, currentSnapshot)) {
                    snapshotFile.createDebugFile(currentSnapshot);

                    List<SnapshotReporter> reporters =
                            snapshotReporters.stream()
                                    .filter(reporter -> reporter.supportsFormat(snapshotSerializer.getOutputFormat()))
                                    .collect(Collectors.toList());

                    if (reporters.isEmpty()) {
                        String comparator = snapshotComparator.getClass().getSimpleName();
                        throw new IllegalStateException(
                                "No compatible reporters found for comparator " + comparator);
                    }

                    List<Throwable> errors = new ArrayList<>();

                    for (SnapshotReporter reporter : reporters) {
                        try {
                            reporter.report(previousSnapshot, currentSnapshot);
                        } catch (Throwable t) {
                            errors.add(t);
                        }
                    }

                    if (!errors.isEmpty()) {
                        throw new SnapshotMatchException("Error(s) matching snapshot(s)", errors);
                    }
                }
            }
        } else {
            if (this.isCI) {
                log.error(
                        "We detected you are running on a CI Server - if this is incorrect please override the isCI() method in SnapshotConfig");
                throw new SnapshotMatchException(
                        "Snapshot ["
                                + resolveSnapshotIdentifier()
                                + "] not found. Has this snapshot been committed ?");
            } else {
                log.warn(
                        "We detected you are running on a developer machine - if this is incorrect please override the isCI() method in SnapshotConfig");
                // Create New Snapshot
                snapshotFile.pushSnapshot(currentSnapshot);
                snapshotFile.pushDebugSnapshot(currentSnapshot);
            }
        }
    }

    private boolean shouldUpdateSnapshot() {
        if (snapshotConfig.updateSnapshot().isPresent() && snapshotConfig.isCI()) {
            throw new SnapshotExtensionException(
                    "isCI=true & update-snapshot="
                            + snapshotConfig.updateSnapshot()
                            + ". Updating snapshots on CI is not allowed");
        }
        if (snapshotConfig.updateSnapshot().isPresent()) {
            return resolveSnapshotIdentifier().contains(snapshotConfig.updateSnapshot().get());
        } else {
            return false;
        }
    }

    private Snapshot getRawSnapshot(Collection<Snapshot> rawSnapshots) {
        synchronized (rawSnapshots) {
            for (Snapshot rawSnapshot : rawSnapshots) {
                if (rawSnapshot.getIdentifier().equals(resolveSnapshotIdentifier())) {
                    return rawSnapshot;
                }
            }
        }
        return null;
    }

    private Snapshot takeSnapshot() {
        SnapshotSerializerContext sg = SnapshotSerializerContext.from(this);
        return snapshotSerializer.apply(current, sg);
    }

    String resolveSnapshotIdentifier() {
        String scenarioFormat = scenario == null ? "" : "[" + scenario + "]";
        return snapshotName() + scenarioFormat;
    }

    private String snapshotName() {
        SnapshotName snapshotName = testMethod.getAnnotation(SnapshotName.class);
        return snapshotName == null
                ? testClass.getName() + "." + testMethod.getName()
                : snapshotName.value();
    }

    void checkValidContext() {
        for (String rw : RESERVED_WORDS) {
            if (snapshotName().contains(rw)) {
                throw new ReservedWordException("snapshot name", rw, RESERVED_WORDS);
            }
            if (scenario != null && scenario.contains(rw)) {
                throw new ReservedWordException("scenario name", rw, RESERVED_WORDS);
            }
        }
    }
}
