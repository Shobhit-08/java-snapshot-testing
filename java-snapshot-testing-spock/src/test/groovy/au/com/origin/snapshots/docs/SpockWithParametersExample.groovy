package au.com.origin.snapshots.docs

import au.com.origin.snapshots.Expect
import au.com.origin.snapshots.annotations.SnapshotName
import au.com.origin.snapshots.spock.EnableSnapshots
import spock.lang.Specification

@EnableSnapshots
class SpockWithParametersExample extends Specification {

    private Expect expect

    @SnapshotName("convert_to_uppercase")
    def 'Convert #scenario to uppercase'() {
        when: 'I convert to uppercase'
        String result = value.toUpperCase();
        then: 'Should convert letters to uppercase'
        // Check you snapshot against your output using a unique scenario
        expect.scenario(scenario).toMatchSnapshot(result)
        where:
        scenario | value
        'letter' | 'a'
        'number' | '1'
    }
}