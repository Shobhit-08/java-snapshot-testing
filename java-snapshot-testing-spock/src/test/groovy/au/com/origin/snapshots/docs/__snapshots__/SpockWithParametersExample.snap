convert_to_uppercase[letter]=[
A
]


convert_to_uppercase[number]=[
1
]